### PR TITLE
Multi node setup for fuzz test

### LINKED ISSUE #4013: Fuzz a pair of state machines, look for nonce mismatch
Some of the last bugs before a release are nonce mismatches.

> hackaugusto 17:03 @jochen started to work on that, we have a few tests here: https://github.com/raiden-network/raiden/tree/master/raiden/tests/fuzz
> IMO the most annoying problem is that currently the state machine checks the signatures
> here: https://github.com/raiden-network/raiden/blob/13d90cabd0eaecc561cf5bf8eedc8ba6a536d406/raiden/transfer/channel.py#L273-L316
> so that means the fuzz either needs to randomly generate valid signatures
> or we need a smartish model of the application

> yoichi 17:04 I see.

> hackaugusto 17:04 this should be improved with this PR: https://github.com/raiden-network/raiden/pull/3860

> so the idea is that all value objects are validated at the boundaries, and that the boundary code is trusted
> to we can just remove the signature verification from the state machine and assume that our own code is correct, and only uses correctness by construction
> after that we have two approaches, either extend the existing asserts
> e.g.: https://github.com/raiden-network/raiden/blob/13d90cabd0eaecc561cf5bf8eedc8ba6a536d406/raiden/transfer/mediated_transfer/mediator.py#L270-L331
> and run the state machine with purely random data
> or build a model of the thing and use AFL or hypothesis
> (AFL would be awesome and I found recently it supports python through a nice little package)

> yoichi 17:07 How can I catch nonse mismatch?
> Maybe I need to couple two state machines, and feed the messages from one into the other?

> hackaugusto 17:12 yes that can work

> just keep in mind that not all nonce mismatches are bugs, so you need synchronous and reliable network model

> IOW, a state change goes into one state machine, events come out, and goes directly to the next state machine

> you will have to do a bit of work to map event to state changes though

> actually, no, an ordered queue is sufficient

> as longs as the events are processed in order things should be fine (by processed I meant that Send events are converted to Receive events for the partner state machine)

### PR DESCRIPTION
**Pull request by jomuel**
Up to now, the test classes for the fuzz test had a single instance of `ChainState`. This PR changes it to a setup with multiple `ChainState` instances representing multiple Raiden nodes that will interact with each other in transfers, so we will not need to mock the interactions. The remaining parts of the fuzz tests will then be both easier to develop and more reliable.

Thanks to @pirapira for suggesting the use of multiple nodes.

Closes #4013

### CODE REVIEW

File: raiden/tests/fuzz/test_state_changes.py [hunk 1]
```diff
@@ -524,136 +584,169 @@ def _new_mediator_transfer(
...
                 block_number=self.block_number + 1,
                 gas_limit=1,
                 block_hash=factories.make_keccak_hash(),
             )
-            result = node.state_transition(self.chain_state, block_state_change)
-            events.extend(result.events)
+            for client in self.address_to_client.values():
+                events = list()
+                result = node.state_transition(client.chain_state, block_state_change)
+                events.extend(result.events)
```
Comment (karlb) [inline raiden/tests/fuzz/test_state_changes.py]: Do we use the `events` variable for anything? If not, we can remove it.

But I must be missing something. To me it looks like the new result is just discarded, so that the `state_transition` doesn't have any effect, either.

File: raiden/tests/fuzz/test_state_changes.py [hunk 2]
```diff
@@ -524,136 +584,169 @@ def _new_mediator_transfer(
...
 
-    @rule(partner=consumes(partners))
-    def settle_channel(self, partner):
-        channel = self.address_to_channel[partner]
+    @rule(address_pair=consumes(address_pairs))
+    def settle_channel(self, address_pair):
+        our_address = address_pair.our_address
+        partner = address_pair.partner_address
+        client = self.address_to_client[our_address]
+        channel = client.address_to_channel[partner]
```
Comment (karlb) [inline raiden/tests/fuzz/test_state_changes.py]: My personal preference would be
```
client = self.address_to_client[address_pair.our_address]
channel = client.address_to_channel[address_pair.partner_address]
```
if the intermediate variables (`our_address` and `partner`) are not used anywhere else in the function and don't add any additional meaning.

File: raiden/tests/fuzz/test_state_changes.py [hunk 3]
```diff
@@ -476,6 +529,12 @@ def update(self, amount, lock):
             )
 
 
+@dataclass
+class WithOurAddress:
+    our_address: typing.Address
+    data: typing.Any
```
Comment (karlb) [inline raiden/tests/fuzz/test_state_changes.py]: This is always an action, isn't it? We could name and type it that way. Or do you think we will use it in more generic cases soon?

**Comment by karlb** [comment]
Did this work without fixing anything in raiden? I would have expected it to detect at least a single unimportant edge case. Are we confident that this actually works as intended?

**Comment by jomuel** [comment]
> Did this work without fixing anything in raiden? I would have expected it to detect at least a single unimportant edge case. Are we confident that this actually works as intended?

We found issues #3000 and #3001 with the fuzz test, and there is a regression test for them at the end of the fuzz test module now.

File: raiden/tests/fuzz/test_state_changes.py (hunk 3 above)
Comment (jomuel) [reply raiden/tests/fuzz/test_state_changes.py]: It will probably always be an action.

File: raiden/tests/fuzz/test_state_changes.py (hunk 2 above)
Comment (jomuel) [reply raiden/tests/fuzz/test_state_changes.py]: True, that looks better

**Comment by karlb** [comment]
> > Did this work without fixing anything in raiden? I would have expected it to detect at least a single unimportant edge case. Are we confident that this actually works as intended?
> 
> We found issues #3000 and #3001 with the fuzz test, and there is a regression test for them at the end of the fuzz test module now.

I would have expected to find something new with this PR. But maybe I'm overly optimistic after having some very positive experiences with a few simple hypothesis tests.

File: raiden/tests/fuzz/test_state_changes.py (hunk 1 above)
Comment (jomuel) [reply raiden/tests/fuzz/test_state_changes.py]: > > > Did this work without fixing anything in raiden? I would have expected it to detect at least a single unimportant edge case. Are we confident that this actually works as intended?
> > 
> > 
> > We found issues #3000 and #3001 with the fuzz test, and there is a regression test for them at the end of the fuzz test module now.
> 
> I would have expected to find something new with this PR. But maybe I'm overly optimistic after having some very positive experiences with a few simple hypothesis tests.

This PR just prepares multiple-node tests, it doesn't test anything that wasn't tested before. There will be follow-up PRs with rules for the remaining parts of the state machine, and I expect them to detect similiar issues again.

Comment (jomuel) [reply raiden/tests/fuzz/test_state_changes.py]: I'll add a TODO that the assertions in this rule are still missing. I just added it in this incomplete form so the rest of the tests run with block events included already.

File: raiden/tests/fuzz/test_state_changes.py [hunk 4]
```diff
@@ -206,85 +229,94 @@ def event(self, description):
...
-                partner_unclaimed + partner_transferred
-                >= self.partner_previous_transferred[address]
-                + self.partner_previous_unclaimed[address]
-            )
-            self.our_previous_transferred[address] = our_transferred
-            self.partner_previous_transferred[address] = partner_transferred
-            self.our_previous_unclaimed[address] = our_unclaimed
-            self.partner_previous_unclaimed[address] = partner_unclaimed
+        for client in self.address_to_client.values():
+            for address, netting_channel in client.address_to_channel.items():
```
[the remark this answers is not in the document]
Comment (jomuel) [reply raiden/tests/fuzz/test_state_changes.py]: I was following the "data classes are just for data" pattern that we use in the core. But we can also do it differently here.

**Comment by jomuel** [comment]
> > > Did this work without fixing anything in raiden? I would have expected it to detect at least a single unimportant edge case. Are we confident that this actually works as intended?
> > 
> > 
> > We found issues #3000 and #3001 with the fuzz test, and there is a regression test for them at the end of the fuzz test module now.
> 
> I would have expected to find something new with this PR. But maybe I'm overly optimistic after having some very positive experiences with a few simple hypothesis tests.

This PR just prepares testing with multiple `ChainState`s. It doesn't test anything that wasn't tested before. There will be follow-ups with rules for the missing parts of the state machine, and I expect them to detect similiar issues again.
